detekt/detekt
# Add detekt compiler plugin to main project

This brings the detekt compiler plugin into the primary project. Aside from bringing the code over, this also makes changes to the Gradle plugin so that the standard plugin as well as the compiler plugin can be used in the same project. This wasn't previously possible due to the naming conflict on the `detekt` extension which was applied by both Gradle plugins.

I'd like to get this merged, then we can work on:
* Confirming the versioning scheme for the compiler plugin which will be sensitive to the Kotlin version the target project is using. I propose just copying KSP's version scheme of `$kotlinVersion-$pluginVersion` e.g. `1.7.20-1.22.0-RC2`
* Docs
* Pushing out at least one release (can then skip step 1 below)
* Work on a fix for `InvalidPackageDeclaration` failures (means we can skip steps 3 & 4 below)

To test:
1. Include these builds in the test project's settings.gradle(.kts) file:
```
includeBuild("../detekt/detekt-gradle-plugin")
includeBuild("../detekt")
```
2. Apply the compiler plugin in a target project:
```kotlin
plugins {
    id("io.github.detekt.gradle.compiler-plugin")
}
```
3. Add a `detekt.yml` config file and disable `InvalidPackageDeclaration` (see [detekt-compiler-plugin/issues/15](https://github.com/detekt/detekt-compiler-plugin/issues/15)):
```yaml
naming:
  InvalidPackageDeclaration:
    active: false
```
4. Set `buildUponDefaultConfig = true`, e.g.:
```kotlin
detekt {
    buildUponDefaultConfig = true
}
```

The compiler plugin cannot be applied to the detekt project itself due to circular dependencies, which I couldn't work around, but hope to some time in future.

Closes #3520
